File: Java/MinStack.java
import java.util.Stack;

public class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int val) {
        stack.push(val);
        if (minStack.isEmpty() || val <= minStack.peek()) {
            minStack.push(val);
        }
    }

    public void pop() {
        if (stack.peek().equals(minStack.peek())) {
            minStack.pop();
        }
        stack.pop();
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }

    public static void main(String[] args) {
        MinStack s = new MinStack();

        s.push(-2);
        s.push(0);
        s.push(4);
        s.push(-3);

        System.out.println(s.getMin()); // -3

        s.pop();
        System.out.println(s.getMin()); // -2

        s.pop();
        System.out.println(s.getMin()); // -2.
        
    }
}

File: Python/MinStack.py
class MinStack:

    def __init__(self):
        self.__stack=[]
        self.__minstack=[]

    def push(self, val: int) -> None:
        self.__stack.append(val)
        if not self.__minstack:
            self.__minstack.append(val)
        else:
            if self.__minstack[-1] >= val:
                self.__minstack.append(val)
            

    def pop(self) -> None:
        if self.__stack[-1]==self.__minstack[-1]:
            self.__minstack.pop()
        self.__stack.pop()

        

    def top(self) -> int:
        return self.__stack[-1]
        

    def getMin(self) -> int:
        return self.__minstack[-1]
        



obj = MinStack()
obj.push(-2)
obj.push(0)
obj.push(4)
obj.push(-3)
print(obj.getMin()) # -3
obj.pop()
print(obj.top()) # 4
print(obj.getMin()) # -2

